M3-5753: Filter deprecated images

cpathipa opened the pull request:
## Description 📝

Goal is to replace manual process of setting flags in DB to get suffix ( "deprecated") added to distro's  in the manager.
- Automatically filters deprecated images 
- Automatically Adds suffix "deprecated"

## Preview 📷

**Before:** 
![image](https://user-images.githubusercontent.com/119517080/205987724-294271b6-854a-4f63-b093-0f2e23300a03.png)

**After:**
![image](https://user-images.githubusercontent.com/119517080/205987855-031a26d2-0e06-4777-bec3-090af6ae3b7f.png)



## How to test 🧪

- Navigate to route /linodes
- Click on create linode
- Validate Images in "Choose a Distribution"

**How do I run relevant unit or e2e tests?**
`yarn test components/ImageSelect/ImageSelect`

hana-linode commented on packages/manager/src/components/ImageSelect/ImageSelect.tsx:
```diff
@@ -16,6 +17,7 @@ import getSelectedOptionFromGroupedOptions from 'src/utilities/getSelectedOption
 import { distroIcons } from './icons';
 import ImageOption from './ImageOption';
 
+const maxEOLFilter = 6;
```
Can we rename this to something like `maxMonthsEOLFilter` so we know what the unit is? And maybe also move this variable to `manager/src/constants.ts`

cpathipa replied:
@hana-linode Yes, I agree, `maxMonthsEOLFilter` is more descriptive. Updated accordingly and moved it to `constants`.

CodyFinn-Akamai commented on packages/manager/src/components/ImageSelect/ImageSelect.tsx:
```diff
@@ -85,10 +87,20 @@ export const imagesToGroupedItems = (images: Image[]) => {
+            .filter((thisImage) => {
+              const { eol } = thisImage;
+              const diff = DateTime.now().diff(DateTime.fromISO(eol!), 'months')
+                .months;
+              return diff > maxEOLFilter && eol !== null ? false : true;
+            })
             .map((thisImage) => {
-              const isDeprecated = thisImage.deprecated;
-              const fullLabel =
-                thisImage.label + (isDeprecated ? ' (Deprecated)' : '');
+              const { eol } = thisImage;
+              const diff = DateTime.fromISO(eol!).diff(DateTime.now(), 'months')
+                .months;
+              let fullLabel = thisImage.label;
+              if (diff <= maxEOLFilter) {
+                fullLabel += ' (Deprecated)';
+              }
```
It is possible to put some of the filtering on the server by using the X-Filter header in the request. `deprecated` is a filterable field. The curl command would look something like this: 

```
curl "https://api.linode.com/v4/images" \
  -H 'X-Filter: { "deprecated": "false" }'
```

hana-linode commented on packages/manager/src/components/ImageSelect/ImageSelect.tsx:
```diff
@@ -85,10 +87,22 @@ export const imagesToGroupedItems = (images: Image[]) => {
+              const diff = DateTime.now().diff(DateTime.fromISO(eol!), 'months')
+                .months;
+              return diff > MAX_MONTHS_EOL_FILTER && eol !== null
+                ? false
+                : true;
+            })
             .map((thisImage) => {
-              const isDeprecated = thisImage.deprecated;
-              const fullLabel =
-                thisImage.label + (isDeprecated ? ' (Deprecated)' : '');
+              const { eol } = thisImage;
+              const diff = DateTime.fromISO(eol!).diff(DateTime.now(), 'months')
```
There seems to be some kind of bug with the time difference, we're showing some images as Deprecated when they haven't reached EOL yet:

- Alpine 3.14 (EOL 2023-05-01)
- Fedora 36 (EOL 2023-05-16)
- Fedora 35 (2023-12-15)
- Ubuntu 18.04 (EOL 2023-04-30)

We should also figure out if we should add Deprecated if the `deprecated` field is true such as for Slackware 14.1

cpathipa replied:
Good catch @hana-linode 

- Above time difference issue was fixed
- Yes, we no longer rely on `deprecated` Since it's a manual process of setting in DB for deprecated images. So, we should rely on `eol` field to dynamically add suffix / filter images that are past end of life.

bnussman-akamai commented on packages/manager/src/components/ImageSelect/ImageSelect.tsx:
```diff
@@ -85,10 +87,22 @@ export const imagesToGroupedItems = (images: Image[]) => {
           label: thisGroup,
           options: group
+            .filter((thisImage) => {
+              const { eol } = thisImage;
+              const diff = DateTime.now().diff(DateTime.fromISO(eol!), 'months')
+                .months;
+              return diff > MAX_MONTHS_EOL_FILTER && eol !== null
+                ? false
+                : true;
```
The non-null assertion (`!`) caught my attention. I found that when `eol` was `null`, `diff` would be `NaN`. Optionally, we could check if it is null before we calculate the time difference. This change is up to you, it results in the same behavior.



```suggestion
              const { eol } = thisImage;
              if (!eol) {
                // no end of life is known so show Image to be safe
                return true;
              }
              const diff = DateTime.now().diff(DateTime.fromISO(eol), 'months')
                .months;

              // if image is past its end of life, hide it, otherwise show it
              return diff > MAX_MONTHS_EOL_FILTER ? false : true;
```

bnussman-akamai reviewed (approved):
Nice job on this logic!

cpathipa commented:
Thank you @hana-linode and @bnussman-akamai for the feedback. Addressed the changes, enhanced unit test coverage and PR is ready for re-review.

CodyFinn-Akamai commented on packages/manager/src/components/ImageSelect/ImageSelect.tsx:
```diff
@@ -85,10 +87,27 @@ export const imagesToGroupedItems = (images: Image[]) => {
+            .filter((thisImage) => {
+              const { eol } = thisImage;
+              if (!eol) {
+                // no end of life is known so show Image to be safe
+                return true;
+              }
+              const diff = DateTime.now().diff(DateTime.fromISO(eol), 'months')
+                .months;
+
+              // if image is past its end of life, hide it, otherwise show it
+              return diff > MAX_MONTHS_EOL_FILTER ? false : true;
+            })
             .map((thisImage) => {
-              const isDeprecated = thisImage.deprecated;
-              const fullLabel =
-                thisImage.label + (isDeprecated ? ' (Deprecated)' : '');
+              const { eol } = thisImage;
+              const diff = DateTime.now().diff(DateTime.fromISO(eol!), 'months')
+                .months;
+              let fullLabel = thisImage.label;
+              // Add suffix 'depricated' to the image if end of life is past
+              if (diff > 0) {
+                fullLabel += ' (Deprecated)';
+              }
```
This could be simplified with a reduce instead of a filter followed by a map

cpathipa replied:
@CodyFinn-Akamai Good call, we can definitely cut down number of operations either with `forEach | reduce`

cpathipa commented on packages/manager/src/components/ImageSelect/ImageSelect.tsx:
```diff
@@ -85,10 +87,27 @@ export const imagesToGroupedItems = (images: Image[]) => {
+                // no end of life is known so show Image to be safe
+                return true;
+              }
+              const diff = DateTime.now().diff(DateTime.fromISO(eol), 'months')
```
sure, renamed `diff` ->  `differenceInMonths`

cpathipa commented on packages/manager/src/components/ImageSelect/ImageSelect.tsx:
```diff
@@ -85,10 +87,27 @@ export const imagesToGroupedItems = (images: Image[]) => {
           label: thisGroup,
           options: group
+            .filter((thisImage) => {
+              const { eol } = thisImage;
```
@hana-linode I like the idea. But, The only downside I see with this is they overridable.

hana-linode replied:
Good point, didn't know they were overridable

cpathipa commented on packages/manager/src/components/ImageSelect/ImageSelect.tsx:
```diff
@@ -85,10 +87,27 @@ export const imagesToGroupedItems = (images: Image[]) => {
+              // if image is past its end of life, hide it, otherwise show it
+              return diff > MAX_MONTHS_EOL_FILTER ? false : true;
+            })
             .map((thisImage) => {
```
@hana-linode The only downside I see is they are overridable. Addressed this within the fun [c8a1955]([c8a1955](https://github.com/linode/manager/pull/8639/commits/c8a195524f9819aa41397f9cf391aef5ed996e82))

cpathipa commented:
@CodyFinn-Akamai and @hana-linode 
Addressed below feedback changes. 
- Replaced map and filter with reduce
- Renamed diff -> differenceInMonths
- Destructuring
